### PR TITLE
Fast method for ElmsBlist when positions are a range with increment 1

### PR DESCRIPTION
**Pull request by stevelinton**
Also used for copying pieces of GF2 vectors around.

### CODE REVIEW

**Comment by ChrisJefferson** [comment]
While I imagine we'd have realised, some nice test which tests a large range of offsets and lengths would be nice, just to make sure there isn't some nasty corner case somewhere.

**Comment by stevelinton** [comment]
@ChrisJefferson I have a stand along C program which is pretty exhaustive. I'll translate it into GAP.

**Comment by fingolfin** [comment]
I added the "not-for-release-notes" label. But perhaps this PR gives a speed boost to something? Then of course this label should be removed again... but I couldn't tell from the description of the PR.

File: src/blister.h [hunk 1]
```diff
@@ -140,6 +140,9 @@ static inline UInt * BLOCKS_BLIST(Obj list)
     return BLOCKS_BLIST_UNSAFE(list);
 }
 
+static inline const UInt * CONST_BLOCKS_BLIST(Obj list) {
+  return (const UInt *)BLOCKS_BLIST(list);
```
Comment (fingolfin) [inline src/blister.h]: I'm afraid this implementation defeats the ultimate purpose: to allow inserting read/write guards in an "optimal" fashion. Really need this (and use 4 spaces for indent):
```C
static inline const UInt * CONST_BLOCKS_BLIST(Obj list)
{
    GAP_ASSERT(IS_BLIST_REP_WITH_COPYING(list));
    return ((const UInt *)(CONST_ADDR_OBJ(list) + 1));
}
```

File: src/blister.h [hunk 2]
```diff
@@ -352,6 +355,147 @@ extern void ConvBlist (
...
+**   starting at bit <to-starting-bit> of *<toblock>.
+**
+**   The full function is provided as an inline function here because it is
+**   used in both blister.c and vecgf2.c and the compiler can do significant
+**   optimisations if some or all of <frombits>, <tobits> and <numbits> are
+**   known at compile time.
+*/
+
+/* constructs a mask that selects bits <from> to <to> inclusive of a UInt */
+static inline UInt mask(UInt from, UInt to)
```
Comment (fingolfin) [inline src/blister.h]: Hm, `mask` is an awfully common name, not sure it's a good idea to use that for a global function.

File: src/blister.h [hunk 3]
```diff
@@ -352,6 +355,147 @@ extern void ConvBlist (
...
+    UInt m = mask(startbit + shift, endbit + shift);
+    *to &= ~m;
+    if (shift >= 0)
+        *to |= ((from << shift) & m);
+    else
+        *to |= ((from >> -shift) & m);
+}
+
+
+static inline  __attribute__((always_inline)) void CopyBits(const UInt * fromblock,
```
Comment (fingolfin) [inline src/blister.h]: Should use `ALWAYS_INLINE` from PR #1779

File: src/vecgf2.c [hunk 4]
```diff
@@ -178,235 +178,22 @@ Obj AddCoeffsGF2VecGF2Vec (
...
-	if (endbits <= cbits)
-	  setlowbits(dptr, endbits, midbits(*sptr++,BIPEB-cbits, endbits));
-	else {
-	  bits = highbits(*sptr++,cbits);
-	  setlowbits(dptr, endbits, bits | (lowbits(*sptr, endbits - cbits) << cbits));
-	}
-      }
-    return;
-  }
+  CopyBits(sptr, soff, dptr, doff, nelts);
```
Comment (fingolfin) [inline src/vecgf2.c]: Have you conducted tests as to whether (and if so: how exactly) this change to the GF2 code affects performance?

**Comment by stevelinton** [comment]
@fingolfin It speeds up ElmsBlist, which was missing this special case, and showed up as 10% of the CPU time in teststandard. More or less equivalent, but much much messier code was already in the vecgf2 applications, so I'd expect little or no change there, but I'll check.

**Comment by fingolfin** [comment]
OK, so it is a performance enhancement. Adjusted labels accordingly.

File: src/blister.h (hunk 3 above)
Comment (fingolfin) [reply src/blister.h]: PR #1779 has been merged, so you can now do this.

File: src/blister.h [hunk 5]
```diff
@@ -352,6 +355,147 @@ extern void ConvBlist (
...
+            CopyInWord(toblock, frombit, BIPEB - 1, *fromblock, 0);
+            fromblock++;
+            toblock++;
+            nbits -= (BIPEB - frombit);
+            frombit = 0;
+            tobit = 0;
+        }
+        /* Now move whole words */
+        if ((wholeblocks = nbits / BIPEB))
+            memcpy((void *)toblock, (void *)fromblock,
```
Comment (fingolfin) [inline src/blister.h]: I don't think you need the typecasts here (and the second one removes the `const` qualifier of `fromblock`.

File: src/blister.h [hunk 6]
```diff
@@ -352,6 +355,147 @@ extern void ConvBlist (
...
+    }
+    /* Main loop for long copies fills whole blocks of destination */
+    UInt m1 = mask(frombit, BIPEB - 1);
+    while (nbits >= BIPEB) {
+        x = (*fromblock++ & m1) >> frombit;
+        x |= (*fromblock & ~m1) << (BIPEB - frombit);
+        *toblock++ = x;
+        nbits -= BIPEB;
+    }
+    /* Finally we may need to fill up a partial block at destination */
```
Comment (fingolfin) [inline src/blister.h]: Why is that comment attached to the preceding loop, instead of the following `if` (to which it refers, doesn't it?)

I.e.: maybe put the empty line before the comment, not after it?

File: src/vecgf2.c [hunk 7]
```diff
@@ -178,235 +178,22 @@ Obj AddCoeffsGF2VecGF2Vec (
...
-    case 60:     dothework(sptr, dptr, 60, dend); break;
-    case 61:     dothework(sptr, dptr, 61, dend); break;
-    case 62:     dothework(sptr, dptr, 62, dend); break;
-    case 63:     dothework(sptr, dptr, 63, dend); break;
-#endif
-    default:  Pr("Illegal shift %i", cbits, 0);
-      SyExit(2);
-    }
+  soff = (smin-1) %BIPEB;
+  doff = (dmin-1) %BIPEB;
```
Comment (fingolfin) [inline src/vecgf2.c]: Either add spaces after `%` or remove the ones before... Or just use `clang-format`?

(If you add spaces, then perhaps also add them around the `/` in the next two lines?)

**Comment by stevelinton** [comment]
Added test and addressed all comments, I think

File: src/blister.h [hunk 8]
```diff
@@ -352,6 +357,148 @@ extern void ConvBlist (
...
+    UInt tailbits;
+    UInt x;
+    UInt wholeblocks;
+    if (!nbits)
+        return;
+    /* Normalize */
+    fromblock += frombit / BIPEB;
+    frombit %= BIPEB;
+    toblock += tobit / BIPEB;
+    tobit %= BIPEB;
```
Comment (fingolfin) [inline src/blister.h]: This function is called from two places, and both already normalize. Is the optimizer clever enough to get rid of the second, redundant normalization?

Otherwise, I'd either remove the normalization in both calling places, or (the approach I personally favor slightly) replace the four lines above by
```C
GAP_ASSERT(frombit < BIPEB);
GAP_ASSERT(tobit < BIPEB);
```

File: src/blister.h [hunk 9]
```diff
@@ -352,6 +357,148 @@ extern void ConvBlist (
 
 /****************************************************************************
 **
+*F  CopyBits( <fromblock>, <from-starting-bit>, <toblock>, <to-starting-bit>,
+**            <numbits> )
+**
+**  `CopyBits' copies <numbits> bits (numbering bits within a UInt
+**   from the least significant to the most significant) starting with
+**   bit number <from-starting-bit> of UInt *<fromblock> to a destination
+**   starting at bit <to-starting-bit> of *<toblock>.
```
Comment (fingolfin) [inline src/blister.h]: Are the two blocks assumed to be non-overlapping? I think it would be a good idea to document this explicitly.

File: src/blister.h (hunk 8 above)
Comment (stevelinton) [reply src/blister.h]: Fine. I put it in because it made it easier to write my C level tests, but you're right it's not needed in GAP.
